Modules/_datetimemodule.c

```diff
@@ -6962,29 +7103,24 @@ _datetime_exec(PyObject *module)
         }
     }
 
-#define CREATE_TYPE(VAR, SPEC, BASE)                    \
-    do {                                                \
-        VAR = (PyTypeObject *)PyType_FromModuleAndSpec( \
-                module, SPEC, (PyObject *)BASE);        \
-        if (VAR == NULL) {                              \
-            goto error;                                 \
-        }                                               \
-    } while (0)
-
-    PyTypeObject *PyDateTime_IsoCalendarDateType = NULL;
-    datetime_state *st = get_datetime_state();
-
-    if (!st->initialized) {
-        CREATE_TYPE(PyDateTime_IsoCalendarDateType, &isocal_spec, &PyTuple_Type);
+    if (old_module != NULL) {
+        datetime_state *st_old = get_module_state(old_module);
+        if (copy_state(st, st_old) < 0) {
+            goto error;
+        }
     }
-#undef CREATE_TYPE
-
-    if (init_state(st, PyDateTime_IsoCalendarDateType) < 0) {
-        goto error;
+    else {
+        if (init_state(st, module) < 0) {
+            goto error;
+        }
     }
 
+    /* For now we only set the objects on the static types once.
+     * We will relax that once each types __dict__ is per-interpreter. */
 #define DATETIME_ADD_MACRO(dict, c, value_expr)         \
     do {                                                \
+      if (PyDict_GetItemString(dict, c) == NULL) {      \
+        assert(!PyErr_Occurred());                      \
         PyObject *value = (value_expr);                 \
```

Comment:
Here's where I've added the special case to avoid creating duplicate objects unnecessarily.  Each use of this macro relates to the `tp_dict` of one of the static types.  There's no need to replace the objects each time the module is loaded since all the objects are effectively immutable.  We'll remove this case when we give the static types as separate `__dict__` per interpreter.